pdns/syncres.hh

```diff
@@ -967,38 +967,53 @@ struct PacketID
 
   bool operator<(const PacketID& b) const
   {
-    int ourSock= tcpsock;
-    int bSock = b.tcpsock;
-    if (tie(remote, ourSock, type) < tie(b.remote, bSock, b.type)) {
+    // We don't want explcit PacketID compare here, but always via predicate classes below
+    assert(0);
+  }
+};
+
+inline ostream& operator<<(ostream & os, const PacketID& pid)
+{
+  return os << "PacketID(id=" << pid.id << ",remote=" << pid.remote.toString() << ",type=" << pid.type << ",tcpsock=" <<
+    pid.tcpsock << ",fd=" << pid.fd << ',' << pid.domain << ')';
+}
+
+inline ostream& operator<<(ostream & os, const shared_ptr<PacketID>& pid)
+{
+  return os << *pid;
+}
+
+struct PacketIDCompare
+{
+  bool operator()(const std::shared_ptr<PacketID>& a, const std::shared_ptr<PacketID>& b) const
+  {
+    if (tie(a->remote, a->tcpsock, a->type) < tie(b->remote, b->tcpsock, b->type)) {
       return true;
     }
-    if (tie(remote, ourSock, type) > tie(b.remote, bSock, b.type)) {
+    if (tie(a->remote, a->tcpsock, a->type) > tie(b->remote, b->tcpsock, b->type)) {
       return false;
     }
 
-    return tie(fd, id, domain) < tie(b.fd, b.id, b.domain);
+    return tie(a->domain, a->fd, a->id) < tie(b->domain, b->fd, b->id);
   }
 };
 
-struct PacketIDBirthdayCompare: public std::binary_function<PacketID, PacketID, bool>
+struct PacketIDBirthdayCompare
```

Comment:
I'm pondering whether adding some comment would be a good idea, something like:
"This comparison needs to be consistent (same ordering, so we can omit right-most keys but not change the order, see CompatibleCompare) with the one in `PacketIDCompare` above."

Reply:
Perhaps some unit tests would help as well :-)

Reply:
will do so.